File: Biome.py
#!/usr/local/bin/python3

# MIT License

# Copyright (c) 2017 Clayton Dorsey
# https://cydy.io @claytondorsey 

import numpy as np
from EmojiGrid import *
import random
from chars import *

def getSectionCords(inputCanvas, sectNum):
    return np.argwhere(inputCanvas == sectNum)

def splitNum(parts, total):
    elements = (np.arange(parts) == np.random.randint(0, parts) for i in range(total))
    return np.sum(elements, axis=0).tolist()

class Biome:

    def replaceSection(self, grid_input, cur, new):
        cords = np.argwhere(grid_input == cur)
        grid_input[cords[:, 0], cords[:, 1]] = new
    
    def crops(self):
        #if random.randint(1, 3) == 1:
        if 1 == 1:
            print('crops: whole')
            crop1 = random.choice(PLANTS_CROPS)
            print('crop:', crop1)

            self.replaceSection(self.sectionCanvas, self.sectNum, crop1)
        pass

    def animals(self):
        placementCords = self.cords
        numberOfCords = placementCords.shape[0] #returns number of cordinates
        fillPercentage = random.uniform(.2, .6)
        numCordsToFill = int(round(numberOfCords * fillPercentage))
        print('filling', numberOfCords, fillPercentage, numCordsToFill)
        placementCords = placementCords[np.random.choice(placementCords.shape[0], numCordsToFill, replace=False), :]

        self.replaceSection(self.sectionCanvas, self.sectNum, ':seedling:') #replace all of section with em-space

        if random.randint(0, 1) == 1:
            print('animals: livestock')
            animalSet = ANIMALS_FARM
        else:
            print('animals: birds')
            animalSet = ANIMALS_BIRDS

        if random.randint(0, 3) == 0: # 3/4 chance one animal biome
            print('animals #: 1')
            animal1 = random.choice(animalSet)
            print('animal1:', animal1)
            for x in placementCords:
                self.sectionCanvas[x[0]][x[1]] = animal1
                pass

        else:
            if 5 > numCordsToFill:
                animalMax = numCordsToFill
            else:
                animalMax = 5
            
            animalNum = random.randint(2, animalMax)
            print('animals #:', animalNum)
            animalChoices = random.sample(animalSet, animalNum)
            animalDistribution = splitNum(animalNum, numCordsToFill)
            #print(animalChoices)
            #print(animalDistribution)
            #print(placementCords)

            z = 0
            for y, animal_output in zip(animalDistribution, animalChoices):
                
                for x in range(y):
                    #print('running ' + str(x) + ' out of', str(y), 'times')
                    #print('placing', animalChoices[animalDistribution.index(y)], 'at',  str(placementCords[z+x][0]), str(placementCords[z+x][1]))
                    self.sectionCanvas[placementCords[z+x][0]][placementCords[z+x][1]] = animal_output
                z += y

        pass

    def field(self):
        placementCords = self.cords
        numberOfCords = placementCords.shape[0] #returns number of cordinates
        fillPercentage = random.uniform(.2, .8)
        numCordsToFill = int(round(numberOfCords * fillPercentage))
        print('filling', numberOfCords, fillPercentage, numCordsToFill)
        placementCords = placementCords[np.random.choice(placementCords.shape[0], numCordsToFill, replace=False), :]

        self.replaceSection(self.sectionCanvas, self.sectNum, ':seedling:') #replace all of section with em-space

        if random.randint(0, 1) == 1:
            print('plants: field')
            plantSet = PLANTS_FIELD
        else:
            print('plants: flowers')
            plantSet = PLANTS_FLOWERS

        if random.randint(0, 3) == 0: # 3/4 chance one plant biome
            print('plants #: 1')
            plant1 = random.choice(plantSet)
            print('plant1:', plant1)
            for x in placementCords:
                self.sectionCanvas[x[0]][x[1]] = plant1
                pass

        else:
            if 5 > numCordsToFill:
                plantMax = numCordsToFill
            else:
                plantMax = 5
            
            plantNum = random.randint(2, plantMax)
            print('plants #:', plantNum)
            plantChoices = random.sample(plantSet, plantNum)
            plantDistribution = splitNum(plantNum, numCordsToFill)
            #print(plantChoices)
            #print(plantDistribution)
            #print(placementCords)

            z = 0
            for y, plant_output in zip(plantDistribution, plantChoices):
                
                for x in range(y):
                    #print('running ' + str(x) + ' out of', str(y), 'times')
                    #print('placing', plantChoices[plantDistribution.index(y)], 'at',  str(placementCords[z+x][0]), str(placementCords[z+x][1]))
                    self.sectionCanvas[placementCords[z+x][0]][placementCords[z+x][1]] = plant_output
                z += y

        pass

    def pond(self):
        placementCords = self.cords
        numberOfCords = placementCords.shape[0] #returns number of cordinates
        fillPercentage = random.uniform(.1, .5)
        numCordsToFill = int(round(numberOfCords * fillPercentage))
        print('filling', numberOfCords, fillPercentage, numCordsToFill)
        placementCords = placementCords[np.random.choice(placementCords.shape[0], numCordsToFill, replace=False), :]

        self.replaceSection(self.sectionCanvas, self.sectNum, ':water_wave:') #replace all of section with em-space

        pawnSet = POND

        if random.randint(0, 3) == 0: # 3/4 chance one pawn biome
            print('pawns #: 1')
            pawn1 = random.choice(pawnSet)
            print('pawn1:', pawn1)
            for x in placementCords:
                self.sectionCanvas[x[0]][x[1]] = pawn1
                pass

        else:
            if 5 > numCordsToFill:
                pawnMax = numCordsToFill
            else:
                pawnMax = 5
            
            pawnNum = random.randint(2, pawnMax)
            print('pawns #:', pawnNum)
            pawnChoices = random.sample(pawnSet, pawnNum)
            pawnDistribution = splitNum(pawnNum, numCordsToFill)
            #print(pawnChoices)
            #print(pawnDistribution)
            #print(placementCords)

            z = 0
            for y, pawn_output in zip(pawnDistribution, pawnChoices):
                
                for x in range(y):
                    #print('running ' + str(x) + ' out of', str(y), 'times')
                    #print('placing', pawnChoices[pawnDistribution.index(y)], 'at',  str(placementCords[z+x][0]), str(placementCords[z+x][1]))
                    self.sectionCanvas[placementCords[z+x][0]][placementCords[z+x][1]] = pawn_output
                z += y

        pass

    def barrier(self):
            #if random.randint(1, 3) == 1:
            if 1 == 1:
                print('barrier: whole')
                barrier1 = random.choice(BARRIERS)
                print('barrier:', barrier1)

                self.replaceSection(self.sectionCanvas, self.sectNum, barrier1)
            pass
    
    def __init__(self, farm_instance, sectionCanvas, sectionNum):
        #super().__init__()
        self.farm = farm_instance
        self.sectionCanvas = sectionCanvas
        self.sectNum = sectionNum
        self.cords = getSectionCords(self.sectionCanvas, self.sectNum)

        print('\nsection number:', str(self.sectNum))

        self.available_biome_types = {0 : self.crops,
                                1 : self.animals,
                                2 : self.field,
                                3 : self.pond}

        if int(self.sectNum) % 2 == 0:
            #print('biome type: BARRIER\n')
            pass
        else:
            self.BIOME_TYPE = random.choice(list(self.available_biome_types.keys()))
            #print('biome type:', self.BIOME_TYPE)
        
        #dimensions = abs(self.cords[-1] - self.cords[0])
	    #sml = np.ones(dimensions + 1)
        #self.canvas = np.full((dimensions + 1), self.sectNum, dtype='U64')
        self.build()
        self.replace()

    def build(self):
        if int(self.sectNum) % 2 == 0:
            print('biome type: BARRIER\n')
            self.barrier()
        else:
            #self.BIOME_TYPE = random.choice(list(self.available_biome_types.keys()))
            #print('biome type:', self.BIOME_TYPE)
            self.available_biome_types[self.BIOME_TYPE]()
        
    def replace(self):
        np.place(self.farm.canvas_strings, self.farm.canvas_strings == self.sectNum, self.sectionCanvas)

    def __str__(self):
        return '\n'.join([''.join([str(elem) for elem in row]) for row in self.sectionCanvas]) + '\n'

    

File: biome.go
//go:build js && wasm

package main

import (
	"fmt"
	"math"
	"math/rand"
)

// Biome represents a section of the farm
type Biome struct {
	grid           *EmojiGrid
	sectionCanvas  [][]string
	sectNum        int
	cords          [][]int
	availableTypes map[int]func(*VerboseLogger)
	rng            *rand.Rand
}

// Helper function to get section coordinates
func getSectionCords(inputCanvas [][]int, sectNum int) [][]int {
	var cords [][]int
	for i, row := range inputCanvas {
		for j, val := range row {
			if val == sectNum {
				cords = append(cords, []int{i, j})
			}
		}
	}
	return cords
}

// Helper function to split numbers using the provided rng
func splitNum(parts, total int, rng *rand.Rand) []int {
	result := make([]int, parts)
	randomIndex := rng.Intn(parts)
	for range total {
		result[randomIndex]++
	}
	return result
}

func NewBiome(grid *EmojiGrid, sectionCanvas [][]string, sectNum int, rng *rand.Rand) *Biome {
	b := &Biome{
		grid:          grid,
		sectionCanvas: sectionCanvas,
		sectNum:       sectNum,
		cords:         getSectionCords(grid.canvasInt, sectNum),
		rng:           rng,
	}

	b.availableTypes = map[int]func(*VerboseLogger){
		0: b.crops,
		1: b.animals,
		2: b.field,
		3: b.pond,
	}

	return b
}

func (b *Biome) replaceSection(gridInput [][]string, cur, new string) {
	for i := range gridInput {
		for j := range gridInput[i] {
			if gridInput[i][j] == cur {
				gridInput[i][j] = new
			}
		}
	}
}

func (b *Biome) crops(logger *VerboseLogger) {
	logger.Log("crops: whole\n")
	crop1 := PLANTS_CROPS[b.rng.Intn(len(PLANTS_CROPS))]
	logger.Log("crop: %s\n", crop1)
	b.replaceSection(b.sectionCanvas, fmt.Sprint(b.sectNum), crop1)
}

func (b *Biome) animals(logger *VerboseLogger) {
	placementCords := b.cords
	numberOfCords := len(placementCords)

	// Fill the entire section with grass background
	for _, cord := range placementCords {
		b.sectionCanvas[cord[0]][cord[1]] = "🌱"
	}

	fillPercentage := b.rng.Float64()*(0.6-0.2) + 0.2
	numCordsToFill := int(math.Round(float64(numberOfCords) * fillPercentage))

	logger.Log("filling %d, %.2f, %d\n", numberOfCords, fillPercentage, numCordsToFill)

	// Shuffle the placement coordinates using b.rng
	b.rng.Shuffle(len(placementCords), func(i, j int) {
		placementCords[i], placementCords[j] = placementCords[j], placementCords[i]
	})

	var animalSet []string
	if b.rng.Intn(2) == 1 {
		logger.Log("animals: livestock\n")
		animalSet = ANIMALS_FARM
	} else {
		logger.Log("animals: birds\n")
		animalSet = ANIMALS_BIRDS
	}

	if b.rng.Intn(4) == 0 { // 1/4 chance one animal biome
		logger.Log("animals #: 1\n")
		animal1 := animalSet[b.rng.Intn(len(animalSet))]
		logger.Log("animal1: %s\n", animal1)
		for i := range numCordsToFill {
			cord := placementCords[i]
			b.sectionCanvas[cord[0]][cord[1]] = animal1
		}
	} else {
		animalMax := min(5, numCordsToFill)
		animalNum := b.rng.Intn(animalMax) + 1
		logger.Log("animals #: %d\n", animalNum)

		animalChoices := randomSample(animalSet, animalNum, b.rng)
		animalDistribution := splitNum(animalNum, numCordsToFill, b.rng)

		z := 0
		for i, animalOutput := range animalChoices {
			for range animalDistribution[i] {
				cord := placementCords[z]
				b.sectionCanvas[cord[0]][cord[1]] = animalOutput
				z++
			}
		}
	}
}

func (b *Biome) field(logger *VerboseLogger) {
	placementCords := b.cords
	numberOfCords := len(placementCords)

	// Fill the entire section with grass background
	for _, cord := range placementCords {
		b.sectionCanvas[cord[0]][cord[1]] = "🌱"
	}

	fillPercentage := b.rng.Float64()*(0.8-0.2) + 0.2
	numCordsToFill := int(math.Round(float64(numberOfCords) * fillPercentage))

	logger.Log("filling %d, %.2f, %d\n", numberOfCords, fillPercentage, numCordsToFill)

	// Shuffle the placement coordinates using b.rng
	b.rng.Shuffle(len(placementCords), func(i, j int) {
		placementCords[i], placementCords[j] = placementCords[j], placementCords[i]
	})

	var plantSet []string
	if b.rng.Intn(2) == 1 {
		logger.Log("plants: field\n")
		plantSet = PLANTS_FIELD
	} else {
		logger.Log("plants: flowers\n")
		plantSet = PLANTS_FLOWERS
	}

	if b.rng.Intn(4) == 0 { // 1/4 chance one plant biome
		logger.Log("plants #: 1\n")
		plant1 := plantSet[b.rng.Intn(len(plantSet))]
		logger.Log("plant1: %s\n", plant1)
		for i := range numCordsToFill {
			cord := placementCords[i]
			b.sectionCanvas[cord[0]][cord[1]] = plant1
		}
	} else {
		plantMax := min(5, numCordsToFill)
		plantNum := b.rng.Intn(plantMax) + 1
		logger.Log("plants #: %d\n", plantNum)

		plantChoices := randomSample(plantSet, plantNum, b.rng)
		plantDistribution := splitNum(plantNum, numCordsToFill, b.rng)

		z := 0
		for i, plantOutput := range plantChoices {
			for range plantDistribution[i] {
				cord := placementCords[z]
				b.sectionCanvas[cord[0]][cord[1]] = plantOutput
				z++
			}
		}
	}
}

// pond builds a pond biome
func (b *Biome) pond(logger *VerboseLogger) {
	logger.Log("Building pond biome\n")
	placementCords := b.cords
	numberOfCords := len(placementCords)
	fillPercentage := b.rng.Float64()*0.4 + 0.1
	numCordsToFill := int(math.Round(float64(numberOfCords) * fillPercentage))
	logger.Log("filling %d, %.2f, %d\n", numberOfCords, fillPercentage, numCordsToFill)

	// Shuffle placementCords using b.rng
	b.rng.Shuffle(len(placementCords), func(i, j int) {
		placementCords[i], placementCords[j] = placementCords[j], placementCords[i]
	})
	placementCords = placementCords[:numCordsToFill]

	b.replaceSection(b.sectionCanvas, fmt.Sprintf("%d", b.sectNum), "🌊")

	pawnSet := POND

	if b.rng.Intn(4) == 0 { // 3/4 chance one pawn biome
		logger.Log("pawns #: 1\n")
		pawn1 := pawnSet[b.rng.Intn(len(pawnSet))]
		logger.Log("pawn1: %s\n", pawn1)
		for _, cord := range placementCords {
			b.sectionCanvas[cord[0]][cord[1]] = pawn1
		}
	} else {
		pawnMax := min(5, numCordsToFill)
		pawnNum := b.rng.Intn(pawnMax) + 1
		logger.Log("pawns #: %d\n", pawnNum)

		pawnChoices := randomSample(pawnSet, pawnNum, b.rng)
		pawnDistribution := splitNum(pawnNum, numCordsToFill, b.rng)

		z := 0
		for i, pawnOutput := range pawnChoices {
			for range pawnDistribution[i] {
				b.sectionCanvas[placementCords[z][0]][placementCords[z][1]] = pawnOutput
				z++
			}
		}
	}
}

// barrier builds a barrier biome
func (b *Biome) barrier(logger *VerboseLogger) {
	logger.Log("Building barrier biome\n")
	logger.Log("barrier: whole\n")
	barrier1 := BARRIERS[b.rng.Intn(len(BARRIERS))]
	logger.Log("barrier: %s\n", barrier1)
	b.replaceSection(b.sectionCanvas, fmt.Sprintf("%d", b.sectNum), barrier1)
}

// Build builds the biome
func (b *Biome) Build(logger *VerboseLogger) {
	logger.Log("section #: %d\n", b.sectNum)

	if b.sectNum%2 == 0 {
		logger.Log("biome type: BARRIER\n")
		b.barrier(logger)
	} else {
		biomeType := b.rng.Intn(len(b.availableTypes))
		b.availableTypes[biomeType](logger)
	}
}

func (b *Biome) Replace() {
	for i, row := range b.grid.canvasStrings {
		for j, val := range row {
			if val == fmt.Sprintf("%d", b.sectNum) {
				b.grid.canvasStrings[i][j] = b.sectionCanvas[i][j]
			}
		}
	}
}

// randomSample selects n random items from a slice without replacement using the provided rng
func randomSample(items []string, n int, rng *rand.Rand) []string {
	if n > len(items) {
		n = len(items)
	}
	result := make([]string, n)
	perm := rng.Perm(len(items))
	for i := range n {
		result[i] = items[perm[i]]
	}
	return result
}
